apm-agent-plugins/apm-logging-plugin/apm-jul-plugin/src/main/java/co/elastic/apm/agent/jul/reformatting/AbstractJulEcsReformattingHelper.java

```diff
@@ -0,0 +1,117 @@
+                                        Formatter originalFormatter) {
+        EcsFormatter ecsFormatter = new EcsFormatter() {
+            @Override
+            protected Map<String, String> getMdcEntries() {
+                // using internal tracer state as ECS formatter is not instrumented within the agent plugin
+                return CorrelationIdMapAdapter.get();
+            }
+        };
+        ecsFormatter.setServiceName(serviceName);
+        ecsFormatter.setServiceVersion(serviceVersion);
+        ecsFormatter.setServiceNodeName(serviceNodeName);
+        ecsFormatter.setEventDataset(eventDataset);
+        if (additionalFields != null && !additionalFields.isEmpty()) {
+            List<AdditionalField> additionalFieldList = new ArrayList<>();
+            for (Map.Entry<String, String> keyValuePair : additionalFields.entrySet()) {
+                additionalFieldList.add(new AdditionalField(keyValuePair.getKey(), keyValuePair.getValue()));
+            }
+            ecsFormatter.setAdditionalFields(additionalFieldList);
+        }
+        ecsFormatter.setIncludeOrigin(false);
+        ecsFormatter.setStackTraceAsArray(false);
+        return ecsFormatter;
+    }
+
+    protected abstract String getShadeFilePatternAndCreateDir() throws IOException;
+
+    @Nullable
+    @Override
+    protected Handler createAndStartEcsAppender(Handler originalHandler, String ecsAppenderName, Formatter ecsFormatter) {
+        StreamHandler shadeHandler = null;
+        if (shouldCreateEcsAppender(originalHandler)) {
+            try {
+                String pattern = getShadeFilePatternAndCreateDir();
+                // In earlier versions, there is only constructor with log file limit given as int, whereas in later ones there are
+                // overloads for both either int or long. Typically, this should be enough, but not necessarily
+                int maxLogFileSize = (int) getMaxLogFileSize();
+                if ((long) maxLogFileSize != getMaxLogFileSize()) {
+                    maxLogFileSize = (int) getDefaultMaxLogFileSize();
+                    oneTimeLogFileLimitWarningLogger.warn("Configured log max size ({} bytes) is too big for JUL settings, which " +
+                        "use int to configure the file size limit. Consider reducing the log max size configuration to a value below " +
+                        "Integer#MAX_VALUE. Defaulting to {} bytes.", getMaxLogFileSize(), maxLogFileSize);
+                }
+                shadeHandler = new FileHandler(pattern, maxLogFileSize, 2, true);
+                shadeHandler.setFormatter(ecsFormatter);
+            } catch (Exception e) {
+                logger.error("Failed to create Log shading FileAppender. Auto ECS reformatting will not work.", e);
+            }
+        }
+        return shadeHandler;
+    }
+
+    protected abstract boolean shouldCreateEcsAppender(Handler originalHandler);
```

Comment:
[minor suggestion]: As someone new to our logging stuff, the name confused me here a little.
I was expecting this to somehow read the `log_ecs_reformatting`, but this actually happens in `createAndMapShadeAppenderFor`.

Is there a reason for not naming this more specific `isFileAppender`? Because these are the two only implementations, right?

Reply:
That's a very good point, I have renamed it to `isFileAppender` for clarity.